File: backend/schemas/session.py
from pydantic import BaseModel, ConfigDict, field_validator
from typing import Optional
from datetime import datetime
import re
from models.session import SessionStatus, SessionMode


class SessionBase(BaseModel):
    session_title: str = "Karaoke"
    session_host: Optional[str] = None
    use_all_songs: bool = True
    allow_song_reuse: bool = False
    session_mode: SessionMode
    songs_per_performer: int = 1
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    changeover_time: Optional[str] = None
    performance_time: Optional[str] = None
    status: SessionStatus = SessionStatus.Active

    @field_validator('start_time', 'end_time', 'changeover_time', 'performance_time')
    @classmethod
    def validate_time_format(cls, v):
        if v is not None and not re.match(r'^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$', v):
            raise ValueError('Time must be in HH:MM format (e.g., "09:30", "14:00")')
        return v


class SessionCreate(SessionBase):
    admin_user_id: int


class SessionUpdate(BaseModel):
    session_title: Optional[str] = None
    session_host: Optional[str] = None
    use_all_songs: Optional[bool] = None
    allow_song_reuse: Optional[bool] = None
    session_mode: Optional[SessionMode] = None
    songs_per_performer: Optional[int] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    changeover_time: Optional[str] = None
    performance_time: Optional[str] = None
    status: Optional[SessionStatus] = None


class SessionResponse(SessionBase):
    session_id: int
    admin_user_id: int
    created_date: datetime
    updated_date: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/apiTypes/session.ts
// Session types matching backend schemas

export enum SessionStatus {
    Active = "Active",
    Completed = "Completed",
    Paused = "Paused"
}

export enum SessionMode {
    Time = 'Time',
    Order = 'Order'
}

export interface Session {
    session_id: number;
    admin_user_id: number;
    session_title: string;
    session_host?: string;
    use_all_songs: boolean;
    allow_song_reuse: boolean;
    session_mode: SessionMode;
    songs_per_performer: number;
    allow_instrument_use: boolean;
    start_time?: string;
    end_time?: string;
    changeover_time?: string;
    performance_time?: string;
    featured_act_name?: string;
    featured_act_start_time?: string;
    featured_act_end_time?: string;
    featured_act_status?: string;
    featured_act_link_url?: string;
    featured_act_link_text?: string;
    custom_link_url?: string;
    custom_link_prompt?: string;
    custom_link_text?: string;
    show_performer_status?: boolean;
    show_song_status?: boolean;
    status: SessionStatus;
    created_date: string;
    updated_date: string;
    song_count?: number;
    performer_count?: number;
    total_session_songs?: number;
}

export interface SessionCreate {
    admin_user_id: number;
    session_title?: string;
    session_host?: string;
    use_all_songs?: boolean;
    allow_song_reuse?: boolean;
    session_mode: SessionMode;
    songs_per_performer?: number;
    allow_instrument_use?: boolean;
    start_time?: string;
    end_time?: string;
    changeover_time?: string;
    performance_time?: string;
    featured_act_name?: string;
    featured_act_start_time?: string;
    featured_act_end_time?: string;
    featured_act_status?: string;
    featured_act_link_url?: string;
    featured_act_link_text?: string;
    custom_link_url?: string;
    custom_link_prompt?: string;
    custom_link_text?: string;
    show_performer_status?: boolean;
    show_song_status?: boolean;
    status?: SessionStatus;
}

export interface SessionUpdate {
    session_title?: string;
    session_host?: string;
    use_all_songs?: boolean;
    allow_song_reuse?: boolean;
    session_mode?: SessionMode;
    songs_per_performer?: number;
    allow_instrument_use?: boolean;
    start_time?: string;
    end_time?: string;
    changeover_time?: string;
    performance_time?: string;
    featured_act_name?: string;
    featured_act_start_time?: string;
    featured_act_end_time?: string;
    featured_act_status?: string;
    featured_act_link_url?: string;
    featured_act_link_text?: string;
    custom_link_url?: string;
    custom_link_prompt?: string;
    custom_link_text?: string;
    show_performer_status?: boolean;
    show_song_status?: boolean;
    status?: SessionStatus;
}
